From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Gregor MacLennan <gmaclennan@gmail.com>
Date: Tue, 28 Jul 2026 19:20:25 +0000
Subject: [PATCH 03/19] build: mobile targets in node.gyp / node.gypi
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

iOS static_library + Android shared library, NODE_MOBILE define, iOS framework deps, and the iOS branch of the node_js2c host-tool condition.

Co-authored-by: Andre Staltz <andre@staltz.com>
Co-authored-by: Julia Samól <j.samol@papers.ch>
Co-authored-by: Jaime Bernardo <jaime@janeasystems.com>
---
 node.gyp  | 39 ++++++++++++++++++++++++++++++++-------
 node.gypi | 22 ++++++++++++++++++----
 2 files changed, 50 insertions(+), 11 deletions(-)

diff --git a/node.gyp b/node.gyp
index 0620850e08..6a5ebfd075 100644
--- a/node.gyp
+++ b/node.gyp
@@ -762,8 +762,13 @@
             }],
           ],
           }, {
-          'sources': [
-            'src/node_snapshot_stub.cc'
+          # nodejs-mobile patch: `conditions`>`not` added to wrap `sources`
+          'conditions': [
+            [ 'not (node_target_type=="static_library" and OS=="ios")', {
+              'sources': [
+                'src/node_snapshot_stub.cc'
+              ],
+            }],
           ],
         }],
         [ 'OS in "linux freebsd openharmony" and '
@@ -937,6 +942,12 @@
             'src/node_snapshot_stub.cc',
           ]
         }],
+        # nodejs-mobile patch:
+        [ 'node_target_type=="static_library" and OS=="ios"', {
+          'sources': [
+            'src/node_snapshot_stub.cc',
+          ]
+        }],
         [ 'node_shared_gtest=="false"', {
           'dependencies': [
             'deps/googletest/googletest.gyp:gtest_prod',
@@ -1010,7 +1021,8 @@
             '<@(node_sqlite_sources)',
           ],
         }],
-        [ 'OS in "linux freebsd mac solaris openharmony" and '
+        # nodejs-mobile patch to mention iOS
+        [ 'OS in "linux freebsd mac ios solaris openharmony" and '
           'target_arch=="x64" and '
           'node_target_type=="executable"', {
           'defines': [ 'NODE_ENABLE_LARGE_CODE_PAGES=1' ],
@@ -1285,6 +1297,12 @@
       'sources': [ '<@(node_cctest_sources)' ],
 
       'conditions': [
+        # nodejs-mobile patch: added this whole `not` block
+        [ 'not (node_target_type=="static_library" and OS=="ios")', {
+          'sources': [
+            'src/node_snapshot_stub.cc',
+          ]
+        }],
         [ 'node_shared_gtest=="false"', {
           'dependencies': [
             'deps/googletest/googletest.gyp:gtest',
@@ -1335,8 +1353,8 @@
         ['OS=="solaris"', {
           'ldflags': [ '-I<(SHARED_INTERMEDIATE_DIR)' ]
         }],
-        # Skip cctest while building shared lib node for Windows
-        [ 'OS=="win" and node_shared=="true"', {
+        # Skip cctest while building shared lib node for Windows and mobile
+        [ 'OS in ("win", "android") and node_shared=="true"', {
           'type': 'none',
         }],
         [ 'node_shared=="true"', {
@@ -1473,7 +1491,8 @@
         [ 'node_shared_libuv=="false"', {
           'dependencies': [ 'deps/uv/uv.gyp:libuv#host' ],
         }],
-        [ 'OS in "linux mac openharmony"', {
+        # nodejs-mobile patch: add ios
+        [ 'OS in "linux mac ios openharmony"', {
           'defines': ['NODE_JS2C_USE_STRING_LITERALS'],
         }],
         [ 'debug_node=="true"', {
@@ -1511,7 +1530,7 @@
       'defines': [ 'NODE_WANT_INTERNALS=1' ],
 
       'sources': [
-        'src/node_snapshot_stub.cc',
+        # nodejs-mobile patch: moved `node_snapshot_stub.cc` to the `not` below
         'tools/snapshot/node_mksnapshot.cc',
       ],
 
@@ -1522,6 +1541,12 @@
       },
 
       'conditions': [
+        # nodejs-mobile patch: added this whole `not` block
+        [ 'not (node_target_type=="static_library" and OS=="ios")', {
+          'sources': [
+            'src/node_snapshot_stub.cc',
+          ]
+        }],
         ['node_write_snapshot_as_array_literals=="true"', {
           'defines': [ 'NODE_MKSNAPSHOT_USE_ARRAY_LITERALS=1' ],
         }],
diff --git a/node.gypi b/node.gypi
index 9e32651032..5178e64dd8 100644
--- a/node.gypi
+++ b/node.gypi
@@ -264,6 +264,11 @@
         'NODE_PLATFORM="darwin"',
       ],
     }],
+    # nodejs-mobile patch: link CoreFoundation and Security frameworks when compiling for iOS
+    [ 'OS=="ios"', {
+      # linking Corefoundation and Security is needed for crypto functionality
+      'libraries': [ '-framework CoreFoundation -framework Security' ],
+    }],
     [ 'OS=="freebsd"', {
       'libraries': [
         '-lutil',
@@ -342,7 +347,7 @@
         ],
       },
     }],
-    [ 'coverage=="true" and node_shared=="false" and OS in "mac freebsd linux openharmony"', {
+    [ 'coverage=="true" and node_shared=="false" and OS in "mac ios freebsd linux openharmony"', {
       'cflags!': [ '-O3' ],
       'ldflags': [ '--coverage',
                    '-g',
@@ -390,11 +395,15 @@
           'defines': [ 'OPENSSL_API_COMPAT=0x10100000L', ],
           'dependencies': [
             './deps/openssl/openssl.gyp:openssl',
-
-            # For tests
-            './deps/openssl/openssl.gyp:openssl-cli',
           ],
           'conditions': [
+            [ 'OS not in "ios android"', {
+              'dependencies': [
+                # Not needed for iOS and Android, doesn't build
+                # For tests
+                './deps/openssl/openssl.gyp:openssl-cli',
+              ],
+            }],
             # -force_load or --whole-archive are not applicable for
             # the static library
             [ 'force_load=="true"', {
@@ -450,5 +459,10 @@
     }, {
       'defines': [ 'HAVE_SQLITE=0' ]
     }],
+    [ 'OS=="android" or OS=="ios"', {
+      'defines': [
+        'NODE_MOBILE',
+      ],
+    }],
   ],
 }
